### ISSUE TITLE
make_fx doesn't work with truly dynamic argument functions (e.g. fx.Interpreter)

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#80903]
### 🐛 Describe the bug

sample which breaks:

```
make_fx(lambda *args: torch.cat(args))([torch.randn(2), torch.randn(2)])
```

Fails with

```
  File "test/test_dynamo_cudagraphs.py", line 145, in cudagraphs
    make_fx(lambda *args: torch.cat(args))([torch.randn(2), torch.randn(2)])
  File "/raid/ezyang/pytorch-scratch2/torch/fx/experimental/proxy_tensor.py", line 281, in wrapped
    t = dispatch_trace(wrap_key(f, args), tracer=fx_tracer, concrete_args=tuple(phs))
  File "/raid/ezyang/pytorch-scratch2/torch/fx/experimental/proxy_tensor.py", line 177, in dispatch_trace
    graph = tracer.trace(root, concrete_args)
  File "/raid/ezyang/torchdynamo/torchdynamo/eval_frame.py", line 88, in _fn
    return fn(*args, **kwargs)
  File "/raid/ezyang/pytorch-scratch2/torch/fx/_symbolic_trace.py", line 667, in trace
    fn, args = self.create_args_for_root(
  File "/raid/ezyang/pytorch-scratch2/torch/fx/_symbolic_trace.py", line 524, in create_args_for_root
    raise RuntimeError(
RuntimeError: Tracing expected 0 arguments but got 1 concrete arguments
```

The problem is FX knows how to see through varargs if there is an inner function its wrapping that has a true argument list, but cannot deal with truly dynamic lists. This is fine for vanilla FX as FX is unwilling to burn in the number of tensors in the list but for AOTAutograd/ProxyTensor we should burn in the list count and trace anyway.

cc @Chillee @zdevito @jamesr66a 

### Versions

master

**Comment by zou3519** [comment]
I'm confused, the function doesn't run without make_fx:
```
f = lambda *args: torch.cat(args)
x = [torch.randn(2), torch.randn(2)]
f(x)
# TypeError: expected Tensor as element 0 in argument 0, but got list
```

**Comment by zou3519** [comment]
Okay the following triggers this as well:
```
f = lambda *args: torch.cat(args)
x = [torch.randn(2), torch.randn(2)]
f(*x)
make_fx(f)(*x)
```
>  but for AOTAutograd/ProxyTensor we should burn in the list count and trace anyway.

Agree

**Comment by ezyang** [comment]
Yeah sorry was a typo.

Here is a really crappy workaround:

```
    # Your interpreter
    t = ApplyCudaGraphs(model)
    
    # Create a wrapper function with the "right" number of arguments to appease
    # FX tracing see https://github.com/pytorch/pytorch/issues/80903
    args = ', '.join([f"v{i}" for i in range(len(inputs))])
    wrapper = eval(f"lambda {args}: t.run({args})", {'t': t})
    
    model = make_fx(wrapper, use_fake=True)(*inputs)
    if model.graph.eliminate_dead_code():
        model.recompile()
```

But @SherlockNoMad informs me that there's another way to do this https://github.com/pytorch/pytorch/blob/master/torch/fx/experimental/proxy_tensor.py#L228

**Comment by zou3519** [comment]
I'm not sure if your question is "make_fx doesn't handle varargs correctly" (we should fix that) or "how do we actually write an interpreter for a graph?".

For the latter, maybe a tangent, but here's how I thought the "interpret the FX graph in eager mode to produce a new graph" would look like:
- User defines a Tensor subclass for the behavior they want (and the Tensor subclass is able to run the "transform" in eager mode).
- Let's say we have a fx graph `g` that we want to transform. We run make_fx over the following function: 
```
def g_prime(*args):
   wrapped_args = wrap_in_subclass(args)
   outputs = g(wrapped_args)
   return unwrap_args(outputs)
```
to generate a graph.

**Comment by ezyang** [comment]
You've identified that there is substantial overlap between fx.Interpreter and Tensor subclass/mode. It seems like they are the same thing. API unification? :/

**Comment by zou3519** [comment]
Yes, we should do API unification. functorch is effectively a stack of interpreters. I don't want there to be three different mechanism to do roughly the same thing: (1) tensor subclasses + modes, (2) functorch transforms, (3) fx.Interpreter

**Comment by ezyang** [comment]
One immediate problem is that fx.Interpreter is a little more powerful than tensor subclass/mode, because it has more "events" (e.g. `call_module` and `get_attr`) which there is no interposition point for in tensor subclasses/modes. Actually, to be super specific, the little bit extra FX has is the monkeypatching it does on Module; otherwise it's implemented with torch function and in principle you can do everything with tensor subclass/mode.

So arguably these are the same thing, but fx.Interpreter is a higher level API than tensor subclasses and modes. Does that mean people should prefer writing fx.Interpreters instead?

**Comment by zou3519** [comment]
Do people want to do module-specific transforms? For the purposes of PT2, the graphs that are obtained are snippets of traces of functions, so if the goal is to "create a common transformation stack that sits in the middle of PT2" we don't need to worry about modules because those will be gone by the time we get here, right?

In general my hot take is that people shouldn't be doing transforms on modules; they should be happening on functions (otherwise we write ourselves into a corner where the transform works "well" on a module but not the variant of the code that doesn't use modules)

**Comment by ezyang** [comment]
It seems that modules are the FX idiomatic way of representing fusion groups and black box function calls. So if I have a straight line AOTAutograd trace and I want to partition it into fusion groups, the way this is currently implemented is to transform the single FX module into a FX modules with submodules representing each fusion group. So I'm not really talking about real nn modules, they are synthetic modules.

**Comment by zou3519** [comment]
> It seems that modules are the FX idiomatic way of representing fusion groups and black box function calls.

Thanks for the clarification. Synthetic modules... don't sound like the right abstraction, I have a lot of questions
